oprf/oprf.go

```diff
@@ -0,0 +1,277 @@
+// Internally, it is a serialized Element.
+type BlindToken []byte
+
+// IssuedToken corresponds to a token that has been issued.
+// Internally, it is a serialized Element.
+type IssuedToken []byte
+
+// Token is the object issuance of the protocol.
+type Token struct {
+	data  []byte
+	blind *group.Scalar
+}
+
+// Evaluation corresponds to the evaluation over a token.
+type Evaluation struct {
+	element []byte
+}
+
+// KeyPair is an struct containing a public and private key.
+type KeyPair struct {
+	pubK  *group.Element
+	PrivK *group.Scalar
+}
+
+// Client is a representation of a Client during protocol execution.
+type Client struct {
+	suite *group.Ciphersuite
+	ctx   []byte
+}
+
+// Server is a representation of a Server during protocol execution.
+type Server struct {
+	suite *group.Ciphersuite
+	ctx   []byte
+	Kp    *KeyPair
+}
+
+func generateCtx(id SuiteID) []byte {
+	ctx := [3]byte{OPRFMode, 0, byte(id)}
+
+	return ctx[:]
+}
+
+// Serialize serializes a KeyPair elements into byte arrays.
+func (kp *KeyPair) Serialize() ([]byte, []byte) {
+	pubK := kp.pubK.Serialize()
+	privK := kp.PrivK.Serialize()
+
+	return pubK, privK
+}
+
+// Deserialize deserializes a KeyPair into an element and field element of the group.
+func (kp *KeyPair) Deserialize(suite *group.Ciphersuite, privK, pubK []byte) error {
+	priv := group.NewScalar(suite)
+	priv.Deserialize(privK)
+
+	pub := group.NewElement(suite)
+	err := pub.Deserialize(pubK)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//GenerateKeyPair generates a KeyPair in accordance with the group.
+func GenerateKeyPair(suite *group.Ciphersuite) *KeyPair {
+	privK := suite.RandomScalar()
+	pubK := suite.ScalarMultBase(privK)
+
+	return &KeyPair{pubK, privK}
+}
+
+func assignKeyPair(suite *group.Ciphersuite, privK, pubK []byte) (*KeyPair, error) {
+	kp := &KeyPair{}
+	err := kp.Deserialize(suite, privK, pubK)
+	if err != nil {
+		return nil, err
+	}
+
+	return kp, nil
+}
+
+func suiteFromID(id SuiteID, ctx []byte) (*group.Ciphersuite, error) {
+	var err error
+	var suite *group.Ciphersuite
+
+	switch id {
+	case OPRFP256:
+		suite, err = group.NewSuite("P-256", ctx)
+	case OPRFP384:
+		suite, err = group.NewSuite("P-384", ctx)
+	case OPRFP521:
+		suite, err = group.NewSuite("P-521", ctx)
+	default:
+		return suite, ErrUnsupportedGroup
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return suite, err
+}
+
+// NewServer creates a new instantiation of a Server. It can create
+// a server with existing keys or use pre-generated keys.
+func NewServer(id SuiteID, privK, pubK []byte) (*Server, error) {
+	ctx := generateCtx(id)
+
+	suite, err := suiteFromID(id, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var keyPair *KeyPair
+	if privK == nil || pubK == nil {
+		keyPair = GenerateKeyPair(suite)
+	} else {
+		keyPair, err = assignKeyPair(suite, privK, pubK)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &Server{
+			suite: suite,
+			ctx:   ctx,
+			Kp:    keyPair},
+		nil
```

Comment:
Can we run `go fmt` on this? (This formatting seems off.)

Reply:
That is the output of `go fmt`.

Reply:
I can put the nil on the same line, so it looks more to what you are accustomed to. But, if the second param to return is on the subsequent line, it looks like this ;)

Reply:
Weird!